netplan_cli/cli/commands/sriov_rebind.py

```diff
@@ -46,5 +61,85 @@ def command_rebind(self):
             if not pcidev.is_pf:
                 logging.warning('{} does not seem to be a SR-IOV physical function'.format(iface))
                 continue
+
+            # There are some hardware-specific configuration that must happen *before* the bind
+            # of VFs to their drivers. Some settings take time to be effective and, when possible,
+            # we need to wait until the driver reports it's ready.
+            self._perform_hardware_specific_quirks(iface, pcidev)
+
             bound_vfs = bind_vfs(pcidev.vfs, pcidev.driver)
             logging.info('{}: bound {} VFs'.format(pcidev, len(bound_vfs)))
+
+    def _perform_hardware_specific_quirks(self, iface: str, pf: PCIDevice):
+        """
+        Perform any hardware-specific quirks for the given SR-IOV device to make
+        sure it's ready before the bind.
+        """
+
+        if pf.driver in ['mlx5_core']:
+            # Mellanox specific quirks
+
+            parser = netplan.Parser()
+            parser.load_yaml_hierarchy(rootdir='/')
+            np_state = netplan.State()
+            np_state.import_parser_results(parser)
+
+            for netdef in np_state.ethernets.values():
+                if netdef.id == iface or netdef.set_name == iface:
+                    if netdef.bond_link:
```

Comment:
I believe there can be some pitfalls in using netplan's state to determine interface's bond state. In the following situation:

* System has two non-bonded interfaces
* User edits netplan config and sets up a bonding
* User calls `netplan apply --sriov-only`

In this scenario, Netplan's state would tell you that the two interfaces are bonded because that's what defined in the config, but since the `--sriov-only` flag [tells netplan to process only sriov stuff and exit](https://github.com/canonical/netplan/blob/d55ae65657d7d0a9611c1be0f6417bff19290fe4/netplan_cli/cli/commands/apply.py#L74), the bond would not be actually created yet.

Reply:
Well, that's true. But one shouldn't run `netplan apply --sriov-only` if they also set up a new bond device in the configuration. In this case you'd need to run a plain `netplan apply`.

But the main reason why I used the netplan state instead of checking the system is that, by the time we do the check, the bond might not be configured yet by networkd. In this case the `master` directory will not be present in sysfs yet. Note that I'm not really sure if this race condition could really happen, but I feel like it safer to use the netplan state.

Reply:
I see, I didn't really consider the possibility of that race condition. I guess both approaches have some drawbacks and it's very likely that user would just call plain `netplan apply` after the change to the config. I just wanted to point out this possibility.

Reply:
Let's keep it open and see if @slyon has a different approach in mind :)